### PR TITLE
bump treewide clap to 4.2.1

### PR DESCRIPTION
**Pull request by fee1-dead**
cc #109302

### CODE REVIEW

File: src/tools/rustfmt/Cargo.toml [hunk 1]
```diff
@@ -37,7 +37,7 @@ annotate-snippets = { version = "0.9", features = ["color"] }
 anyhow = "1.0"
 bytecount = "0.6"
 cargo_metadata = "0.14"
-clap = { version = "3.1", features = ["derive"] }
+clap = { version = "4.2.1", features = ["derive"] }
```
Comment (Mark-Simulacrum) [inline src/tools/rustfmt/Cargo.toml:40]: @rust-lang/rustfmt Are you OK with this landing upstream? It seems like there's some changes to code here as well so maybe we should go through PR to rustfmt?

**Review by Mark-Simulacrum — APPROVED** [review]
r=me modulo rustfmt ping

File: src/tools/rustfmt/Cargo.toml (hunk 1 above)
Comment (ytmimi) [reply src/tools/rustfmt/Cargo.toml:40]: Adding clap 3.1 gave us some trouble https://github.com/rust-lang/rustfmt/issues/5395, but I think bumping to 4.2.1 upstream should be fine. I've gone ahead and tested the 4.2.1 changes locally. There are some slight changes in the `--help` output, but I also think that's fine. @calebcartwright what are your thoughts?

## 3.1
Screen Shot 2023-04-09 at 12 43 26 PM

## 4.2.1
Screen Shot 2023-04-09 at 12 45 03 PM

Comment (calebcartwright) [reply src/tools/rustfmt/Cargo.toml:40]: In general I don't think subtrees should be updated in tree unless absolutely necessary (e.g. rustc_ast updates requiring corresponding updates in clippy, rustfmt, etc.). I understand the benefit of making a broad update like this in tree, but I don't particularly like it (we run a reduced test suite here, lockfile considerations can complicate subtree syncs, etc.)

Suppose we can let this one slide (thanks Yacin for taking a closer look), but in the future would really want to see something like this come directly to the rustfmt repo and then make its way in-tree here via the usual sync process

**Comment by fee1-dead** [comment]
Thanks for the review. I will send PR to the subtree next time changes like this are to be made.

@bors r=Mark-Simulacrum

**Comment by fee1-dead** [comment]
@bors r-

failed in rollup

**Comment by fee1-dead** [comment]
Fixed the error by adding the feature in `rustc_workspace_hack`

@bors r=Mark-Simulacrum

**Comment by compiler-errors** [comment]
> https://github.com/rust-lang/rust/pull/110077#issuecomment-1504317967

Thanks. I was suspicious this was the reason why it failed but couldn't tell.

**Comment by fee1-dead** [comment]
We should wait for #109133 to be merged first such that this is easier. In the meantime perhaps I could send the rustfmt part as a PR to rustfmt first.

**Comment by Dylan-DPC** [comment]
@fee1-dead #109133 is merged now which unblocks this.

**Comment by fee1-dead** [comment]
This is still blocked on https://github.com/rust-lang/rustfmt/pull/5749 and a submodule sync.

**Comment by fee1-dead** [comment]
Closing, as it looks like clap has already been bumped.

**Pull request closed without merging**